File: inference/fhe_layers/par_lower_diag_pcmm.cpp
/*
 * Copyright (c) 2025-2026 CipherFlow (Shenzhen) Co., Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

#include "par_lower_diag_pcmm.h"
#include "layer_util.h"
#include <algorithm>
#include <cassert>
#include <utility>

using namespace std;
using namespace lattisense;

ParLowerDiagPCMM::ParLowerDiagPCMM(const CkksParameter& param_in,
                                   const Duo& shape_X_T,
                                   uint32_t n_heads,
                                   uint32_t head_dim,
                                   const Array<double, 2>& W_mat,
                                   uint32_t level_X,
                                   Array<double, 1>&& bias)
    : Layer(param_in) {
    assert(level_X >= 2);
    level_ = level_X;
    H_prepad_ = n_heads;
    m_ = head_dim;
    assert(H_prepad_ > 0);
    assert(m_ > 0 && (m_ & (m_ - 1)) == 0);

    d_prepad_ = H_prepad_ * m_;
    in_rows_ = shape_X_T[0];
    n_prepad_ = shape_X_T[1];

    H_ = next_pow2(H_prepad_);
    n_ = next_pow2(n_prepad_);
    d_ = H_ * m_;
    assert(n_ >= m_);
    assert(n_ % m_ == 0);

    n_slot_ = param_.get_n() / 2;
    segment_len_ = H_ * n_;
    assert(n_slot_ % segment_len_ == 0);
    c_ = n_slot_ / segment_len_;
    assert(m_ % c_ == 0);
    m_c_ = m_ / c_;

    W_T_rows_ = W_mat.get_shape()[1];
    W_T_cols_ = W_mat.get_shape()[0];
    assert(in_rows_ == W_T_cols_);
    out_cols_ = W_T_rows_;

    K_row_ = div_ceil(W_T_rows_, d_prepad_);
    K_col_ = div_ceil(W_T_cols_, d_prepad_);
    if (K_row_ == K_col_) {
        mode_ = Mode::SQUARE;
        assert(K_row_ == 1);
    } else if (K_row_ > K_col_) {
        mode_ = Mode::EXPAND;
        assert(K_col_ == 1);
    } else {
        mode_ = Mode::REDUCE;
        assert(K_row_ == 1);
    }
    K_ = std::max(K_row_, K_col_);

    // weight matrix is tranposed here
    W_padded_.resize(K_);
    for (uint32_t mb = 0; mb < K_; mb++) {
        Array<double, 2> W_sub({d_, d_});
        for (uint32_t row = 0; row < d_prepad_; row++) {
            for (uint32_t col = 0; col < d_prepad_; col++) {
                uint32_t src_row = (mode_ == Mode::EXPAND) ? mb * d_prepad_ + row : row;
                uint32_t src_col = (mode_ == Mode::REDUCE) ? mb * d_prepad_ + col : col;
                double val = 0.0;
                if (src_row < W_T_rows_ && src_col < W_T_cols_) {
                    val = W_mat.get(src_col, src_row);
                }
                W_sub.set(row, col, val);
            }
        }
        W_padded_[mb] = std::move(W_sub);
    }

    if (bias.get_size() > 0) {
        has_bias_ = true;
        assert(bias.get_size() == out_cols_);
        bias_vals_ = std::move(bias);
    }
}

void ParLowerDiagPCMM::prepare_weight() {
    CkksContext ctx = CkksContext::create_empty_context(param_);
    double pt_scale = param_.get_q(level_);
    double mask_scale = param_.get_q(level_ - 1);
    double bias_scale = param_.get_default_scale();

    pt_A_.clear();
    pt_A_.resize(K_);
    for (uint32_t mb = 0; mb < K_; mb++) {
        pt_A_[mb].resize(H_);
        for (uint32_t i = 0; i < H_; i++) {
            pt_A_[mb][i].reserve(m_c_ * c_ * m_c_);
            for (uint32_t j = 0; j < m_c_; j++) {
                for (uint32_t ell = 0; ell < c_; ell++) {
                    for (uint32_t r = 0; r < m_c_; r++) {
                        vector<double> plaintext(n_slot_, 0.0);
                        for (uint32_t tau = 0; tau < c_; tau++) {
                            uint32_t b_diag_idx = c_ * j + ((tau + ell) % c_);
                            uint32_t out_diag_idx = c_ * r + tau;
                            uint32_t segment_base = tau * segment_len_;
                            uint32_t rotate_step = (out_diag_idx * H_) % segment_len_;
                            uint32_t diag_idx = (b_diag_idx + m_ - (out_diag_idx % m_)) % m_;
                            for (uint32_t t = 0; t < n_; t++) {
                                for (uint32_t h = 0; h < H_; h++) {
                                    uint32_t block_row_base = ((i + h) % H_) * m_;
                                    uint32_t block_col_base = h * m_;
                                    uint32_t row = t % m_;
                                    uint32_t col = (diag_idx + row) % m_;
                                    uint32_t src_row = block_row_base + row;
                                    uint32_t src_col = block_col_base + col;
                                    uint32_t slot_in_segment = t * H_ + h;
                                    uint32_t rotated_slot =
                                        (slot_in_segment + segment_len_ - rotate_step) % segment_len_;
                                    plaintext[segment_base + rotated_slot] = W_padded_[mb].get(src_row, src_col);
                                }
                            }
                        }
                        pt_A_[mb][i].push_back(ctx.encode_ringt(plaintext, pt_scale));
                    }
                }
            }
        }
    }

    mask_wrap_pt_.clear();
    mask_wrap_pt_.resize(H_);
    for (uint32_t i = 1; i < H_; i++) {
        vector<double> mask(n_slot_, 0.0);
        for (uint32_t segment_start = 0; segment_start < n_slot_; segment_start += segment_len_) {
            for (uint32_t t = 0; t < n_; t++) {
                uint32_t group_start = segment_start + t * H_;
                for (uint32_t h = H_ - i; h < H_; h++) {
                    mask[group_start + h] = 1.0;
                }
            }
        }
        mask_wrap_pt_[i] = ctx.encode_ringt(mask, mask_scale);
    }

    bias_pt_.clear();
    if (has_bias_) {
        uint32_t output_mbs = (mode_ == Mode::EXPAND) ? K_ : 1;
        bias_pt_.resize(output_mbs * m_c_);
        for (uint32_t mb = 0; mb < output_mbs; mb++) {
            for (uint32_t r = 0; r < m_c_; r++) {
                vector<double> bias_vec(n_slot_, 0.0);
                for (uint32_t local_diag = 0; local_diag < c_; local_diag++) {
                    uint32_t diag_idx = r * c_ + local_diag;
                    uint32_t segment_base = local_diag * segment_len_;
                    for (uint32_t t = 0; t < n_; t++) {
                        for (uint32_t h = 0; h < H_; h++) {
                            uint32_t out_row = mb * d_prepad_ + h * m_ + ((diag_idx + t) % m_);
                            if (h < H_prepad_ && out_row < out_cols_) {
                                bias_vec[segment_base + t * H_ + h] = bias_vals_.get(out_row);
                            }
                        }
                    }
                }
                bias_pt_[mb * m_c_ + r] = ctx.encode_ringt(bias_vec, bias_scale);
            }
        }
    }
}

std::vector<CkksCiphertext> ParLowerDiagPCMM::run_core(CkksContext& ctx,
                                                       const std::vector<CkksCiphertext>& input_cts,
                                                       const std::vector<uint32_t>& mb_indices) const {
    double default_scale = param_.get_default_scale();
    vector<CkksCiphertext> reduced(m_c_);
    bool reduced_init = false;

    for (uint32_t local_mb = 0; local_mb < mb_indices.size(); local_mb++) {
        uint32_t weight_mb = mb_indices[local_mb];
        uint32_t input_offset = local_mb * m_c_;

        vector<vector<CkksCiphertext>> ct_Br(m_c_);
        for (uint32_t j = 0; j < m_c_; j++) {
            ct_Br[j].resize(c_);
            ct_Br[j][0] = input_cts[input_offset + j].copy();
            for (uint32_t ell = 1; ell < c_; ell++) {
                ct_Br[j][ell] = ctx.rotate(input_cts[input_offset + j], (int)(segment_len_ * ell));
            }
        }

        vector<vector<CkksCiphertext>> ct_ir(H_);
        for (uint32_t i = 0; i < H_; i++) {
            ct_ir[i].resize(m_c_);
            for (uint32_t r = 0; r < m_c_; r++) {
                CkksCiphertext acc(0);
                bool init = false;
                for (uint32_t j = 0; j < m_c_; j++) {
                    for (uint32_t ell = 0; ell < c_; ell++) {
                        uint32_t pt_idx = (j * c_ + ell) * m_c_ + r;
                        auto pt_mul = ctx.ringt_to_mul(pt_A_[weight_mb][i][pt_idx], level_);
                        auto product = ctx.mult_plain_mul(ct_Br[j][ell], pt_mul);
                        if (!init) {
                            acc = std::move(product);
                            init = true;
                        } else {
                            acc = ctx.add(acc, product);
                        }
                    }
                }
                ct_ir[i][r] = ctx.rescale(acc, default_scale);
            }
        }

        vector<CkksCiphertext> ct_C(m_c_);
        for (uint32_t r = 0; r < m_c_; r++) {
            CkksCiphertext acc = ctx.drop_level(ct_ir[0][r]);
            for (uint32_t i = 1; i < H_; i++) {
                auto mask_mul = ctx.ringt_to_mul(mask_wrap_pt_[i], level_ - 1);
                auto ct_R = ctx.rescale(ctx.mult_plain_mul(ct_ir[i][r], mask_mul), default_scale);
                auto ct_i_drop = ctx.drop_level(ct_ir[i][r]);
                auto ct_L = ctx.sub(ct_i_drop, ct_R);
                auto ct_R_rot = ctx.rotate(ct_R, (int)(H_ - i));
                auto ct_L_rot = ctx.rotate(ct_L, -(int)i);
                auto ct_prime = ctx.add(ct_R_rot, ct_L_rot);
                acc = ctx.add(acc, ct_prime);
            }
            ct_C[r] = std::move(acc);
        }

        if (!reduced_init) {
            reduced = std::move(ct_C);
            reduced_init = true;
        } else {
            for (uint32_t r = 0; r < m_c_; r++) {
                reduced[r] = ctx.add(reduced[r], ct_C[r]);
            }
        }
    }

    return reduced;
}

FeatureMatEncrypted ParLowerDiagPCMM::run(CkksContext& ctx, const FeatureMatEncrypted& X_T) {
    assert(X_T.level == level_);
    assert(X_T.shape[0] == n_prepad_);
    assert(X_T.shape[1] == in_rows_);
    assert(X_T.head_shape[0] == n_prepad_);
    assert(X_T.head_shape[1] == m_);
    assert(X_T.matmul_block_size == m_);
    assert(X_T.data.size() == K_col_ * m_c_);

    FeatureMatEncrypted result(&ctx, X_T.level);
    result.level = X_T.level - 2;
    result.shape = {n_prepad_, out_cols_};
    result.head_shape = {n_prepad_, m_};
    result.matmul_block_size = m_;

    if (mode_ == Mode::EXPAND) {
        for (uint32_t mb = 0; mb < K_; mb++) {
            auto mb_cts = run_core(ctx, X_T.data, {mb});
            if (has_bias_) {
                for (uint32_t r = 0; r < m_c_; r++) {
                    mb_cts[r] = ctx.add_plain_ringt(mb_cts[r], bias_pt_[mb * m_c_ + r]);
                }
            }
            for (auto& ct : mb_cts) {
                result.data.push_back(std::move(ct));
            }
        }
    } else {
        vector<uint32_t> all_mbs(K_);
        for (uint32_t i = 0; i < K_; i++) {
            all_mbs[i] = i;
        }
        result.data = run_core(ctx, X_T.data, all_mbs);
        if (has_bias_) {
            for (uint32_t r = 0; r < m_c_; r++) {
                result.data[r] = ctx.add_plain_ringt(result.data[r], bias_pt_[r]);
            }
        }
    }

    return result;
}

File: inference/model_generator/layers/par_lower_diag_pcmm.py
# Copyright (c) 2025-2026 CipherFlow (Shenzhen) Co., Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

import math
import sys
from pathlib import Path
from collections import defaultdict

sys.path.insert(0, str(Path(__file__).resolve().parents[3]))

from inference.lattisense.frontend.custom_task import *
from inference.model_generator.layers.fhe_op_utils import memory_from_pt_counts


op_class = 'ParLowerDiagPCMM'


def _next_pow2(n: int) -> int:
    p = 1
    while p < n:
        p <<= 1
    return p


class ParLowerDiagPCMM:
    """Python model-generator counterpart of C++ ParLowerDiagPCMM."""

    def __init__(
        self,
        shape_X_T: tuple,
        W_T_shape: tuple,
        n_heads: int,
        head_dim: int,
        n_slot: int,
        has_bias: bool = False,
    ):
        self.H_prepad = n_heads
        self.m = head_dim
        assert self.H_prepad > 0
        assert self.m > 0 and (self.m & (self.m - 1)) == 0

        self.d_prepad = self.H_prepad * self.m
        self.in_rows = shape_X_T[0]
        self.n_prepad = shape_X_T[1]

        self.H = _next_pow2(self.H_prepad)
        self.n = _next_pow2(self.n_prepad)
        self.d = self.H * self.m
        assert self.n >= self.m
        assert self.n % self.m == 0

        self.n_slot = n_slot
        self.segment_len = self.H * self.n
        assert n_slot % self.segment_len == 0
        self.c = n_slot // self.segment_len
        assert self.m % self.c == 0
        self.m_c = self.m // self.c

        self.W_T_rows = W_T_shape[0]
        self.W_T_cols = W_T_shape[1]
        assert self.in_rows == self.W_T_cols
        self.out_rows = self.W_T_rows

        self.K_row = math.ceil(self.W_T_rows / self.d_prepad)
        self.K_col = math.ceil(self.W_T_cols / self.d_prepad)
        if self.K_row == self.K_col:
            self.mode = 'SQUARE'
            assert self.K_row == 1
        elif self.K_row > self.K_col:
            self.mode = 'EXPAND'
            assert self.K_col == 1
        else:
            self.mode = 'REDUCE'
            assert self.K_row == 1
        self.K = max(self.K_row, self.K_col)
        self.has_bias = has_bias
        self.n_out_mbs = self.K if self.mode == 'EXPAND' else 1

    def _run_core(self, input_cts: list, mb_indices: list, pt_A: list, mask_wrap_pt: list) -> list:
        reduced = [None] * self.m_c

        for local_mb, weight_mb in enumerate(mb_indices):
            input_offset = local_mb * self.m_c

            ct_Br = []
            for j in range(self.m_c):
                row = [input_cts[input_offset + j]]
                for ell in range(1, self.c):
                    row.append(rotate_cols(input_cts[input_offset + j], [self.segment_len * ell])[0])
                ct_Br.append(row)

            ct_ir = []
            for i in range(self.H):
                row_i = []
                for r in range(self.m_c):
                    acc = None
                    for j in range(self.m_c):
                        for ell in range(self.c):
                            pt_idx = (j * self.c + ell) * self.m_c + r
                            product = mult(ct_Br[j][ell], pt_A[weight_mb][i][pt_idx])
                            acc = product if acc is None else add(acc, product)
                    assert acc is not None
                    row_i.append(rescale(acc))
                ct_ir.append(row_i)

            ct_C = []
            for r in range(self.m_c):
                acc = drop_level(ct_ir[0][r])
                for i_idx in range(self.H - 1):
                    i = i_idx + 1
                    ct_R = rescale(mult(ct_ir[i][r], mask_wrap_pt[i_idx]))
                    ct_i_drop = drop_level(ct_ir[i][r])
                    ct_L = sub(ct_i_drop, ct_R)
                    ct_R_rot = rotate_cols(ct_R, [self.H - i])[0]
                    ct_L_rot = rotate_cols(ct_L, [-i])[0]
                    acc = add(acc, add(ct_R_rot, ct_L_rot))
                ct_C.append(acc)

            if reduced[0] is None:
                reduced = ct_C
            else:
                reduced = [add(reduced[r], ct_C[r]) for r in range(self.m_c)]

        return reduced

    def call(self, input_cts: list, pt_A: list, mask_wrap_pt: list, bias_pt: list | None = None) -> list:
        if self.mode == 'EXPAND':
            result = []
            for mb in range(self.K):
                mb_cts = self._run_core(input_cts, [mb], pt_A, mask_wrap_pt)
                if bias_pt is not None:
                    mb_cts = [add(ct, bias_pt[mb * self.m_c + r]) for r, ct in enumerate(mb_cts)]
                result.extend(mb_cts)
            return result

        result = self._run_core(input_cts, list(range(self.K)), pt_A, mask_wrap_pt)
        if bias_pt is not None:
            result = [add(ct, bias_pt[r]) for r, ct in enumerate(result)]
        return result

    def call_custom_compute(self, input_cts: list, data_source) -> list:
        pt_A = []
        for mb in range(self.K):
            pt_mb = []
            for i in range(self.H):
                pt_i = []
                for j in range(self.m_c):
                    for ell in range(self.c):
                        for r in range(self.m_c):
                            node = CkksPlaintextRingtNode(f'encode_pt_A_{mb}_{i}_{j}_{ell}_{r}')
                            custom_compute(
                                inputs=[data_source],
                                output=node,
                                type='encode_pt',
                                attributes={
                                    'op_class': op_class,
                                    'type': 'pt_A',
                                    'mb': mb,
                                    'i': i,
                                    'j': j,
                                    'ell': ell,
                                    'r': r,
                                },
                            )
                            pt_i.append(node)
                pt_mb.append(pt_i)
            pt_A.append(pt_mb)

        mask_wrap_pt = []
        for i_idx in range(self.H - 1):
            node = CkksPlaintextRingtNode(f'encode_pt_mask_wrap_{i_idx}')
            custom_compute(
                inputs=[data_source],
                output=node,
                type='encode_pt',
                attributes={'op_class': op_class, 'type': 'mask_wrap_pt', 'i_idx': i_idx},
            )
            mask_wrap_pt.append(node)

        bias_pt = None
        if self.has_bias:
            bias_pt = []
            for mb in range(self.n_out_mbs):
                for r in range(self.m_c):
                    node = CkksPlaintextRingtNode(f'encode_pt_bias_{mb}_{r}')
                    custom_compute(
                        inputs=[data_source],
                        output=node,
                        type='encode_pt',
                        attributes={'op_class': op_class, 'type': 'bias_pt', 'mb': mb, 'r': r},
                    )
                    bias_pt.append(node)

        return self.call(input_cts, pt_A, mask_wrap_pt, bias_pt)

    def get_memory(self, bytes_per_plaintext: int = 0) -> dict[str, int]:
        bias_count = self.n_out_mbs * self.m_c if self.has_bias else 0
        counts = {
            'weight': self.K * self.H * self.m_c * self.m,
            'bias': bias_count,
            'mask': self.H - 1,
        }
        return memory_from_pt_counts(counts, bytes_per_plaintext)

    def get_fhe_op_count(self, level: int) -> dict:
        """Count FHE primitive operations grouped by level.

        The count follows ``_run_core``: input replication rotations and
        plaintext matrix products happen at the input level; wrap masking
        consumes one more level; rotations/additions that route wrapped pieces
        are counted at the post-mask level.
        """
        ops = defaultdict(lambda: {'rotate': 0, 'mult_plain': 0, 'mult': 0, 'add': 0, 'rescale': 0})

        n_processed_mbs = self.K
        n_core_calls = self.K if self.mode == 'EXPAND' else 1
        lv = level

        # Build ct_Br and ct_ir for each processed weight megablock.
        ops[lv]['rotate'] += n_processed_mbs * self.m_c * max(0, self.c - 1)
        ops[lv]['mult_plain'] += n_processed_mbs * self.H * self.m_c * self.m
        ops[lv]['add'] += n_processed_mbs * self.H * self.m_c * max(0, self.m - 1)
        ops[lv]['rescale'] += n_processed_mbs * self.H * self.m_c

        # Apply wrap masks to nonzero head offsets.
        wrap_terms = n_processed_mbs * self.m_c * max(0, self.H - 1)
        ops[lv - 1]['mult_plain'] += wrap_terms
        ops[lv - 1]['rescale'] += wrap_terms

        # Route wrapped pieces and accumulate within each output ciphertext.
        ops[lv - 2]['rotate'] += wrap_terms * 2
        ops[lv - 2]['add'] += wrap_terms * 3

        # REDUCE mode combines multiple processed megablocks in one core call.
        if self.mode != 'EXPAND':
            ops[lv - 2]['add'] += max(0, self.K - 1) * self.m_c

        if self.has_bias:
            ops[lv - 2]['add'] += n_core_calls * self.m_c

        return dict(ops)
